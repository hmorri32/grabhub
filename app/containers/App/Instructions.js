import React from 'react';

const Instructions = () => (
  <article>
    <h1>Instructions</h1>
    <h2>About HCW Grabhub</h2>
    <p>
      Grabhub was generated using{' '}
      <a href="https://www.reactboilerplate.com/" target="_blank">
        React Boilerplate
      </a>
      , a boilerplate application developed by Max Stoiber for building scalable
      React Web applications. Please get started by familiarizing yourself with
      the project if you are not already. The boilerplate ships with an example
      app, an interface for fetching links to a Github user&#39;s repos.
      Normally when you use React Boilerplate, You remove the sample app by
      running <code>npm run clean</code> before beginning development. For
      Grabhub, we have kept the sample app and would like you to add another
      feature.
    </p>
    <h2>The Challenge</h2>
    <h3>The client has requested a new feature...</h3>
    <blockquote>
      <em>
        As a person, when I visit the home page, I see a field to enter a Github
        username. When I enter a username and hit enter, add a link to a user
        detail page above the list of repos that returns as a search result.
        This link would take the person to a new page that provides the
        following information: the user&#39;s name, avatar image, a link to
        their github profile, when they joined the system, and a list of their
        followers.
      </em>
    </blockquote>
    <p>
      Follow the established pattern to add the new feature. Maintain the 100%
      test coverage established in the repo.
    </p>

    <h2>Resources</h2>
    <ul>
      <li>
        Try <code>curl https://api.github.com/users/__username__</code> from
        your terminal to get details about a github user.
      </li>
      <li>
        The boilerplate&#39;s{' '}
        <a
          target="_blank"
          href="https://github.com/HaughtCodeworks/front-end-technical-challenge#the-react-boilerplate-readme-follows"
        >
          readme
        </a>{' '}
        has been left largely intact and is packed with info about using the
        boilerplate.
      </li>
      <li>
        Feel free to add any libraries or tools you like to aid in development.
      </li>
      <li>
        Feel free to reach out with questions to{' '}
        <a href="mailto:jeneve.parrish@haughtcodeworks.com">Jeneve.</a>
      </li>
    </ul>
  </article>
);

export default Instructions;
