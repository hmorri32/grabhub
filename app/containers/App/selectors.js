import { createSelector } from 'reselect';

const selectGlobal = state => state.get('global');
const selectRouter = state => state.get('router');

const makeSelectCurrentUser = () =>
  createSelector(selectGlobal, globalState => globalState.get('currentUser'));

const makeSelectLoading = () =>
  createSelector(selectGlobal, globalState => globalState.get('loading'));

const makeSelectError = () =>
  createSelector(selectGlobal, globalState => globalState.get('error'));

const makeSelectRepos = () =>
  createSelector(selectGlobal, globalState =>
    globalState.getIn(['userData', 'repositories']),
  );

const makeSelectLocation = () =>
  createSelector(selectRouter, routerState =>
    routerState.get('location').toJS(),
  );

const makeSelectUserProfile = () =>
  createSelector(selectGlobal, globalState =>
    globalState.getIn(['userData', 'profile']),
  );

const makeSelectUserFollowers = () =>
  createSelector(selectGlobal, globalState =>
    globalState.getIn(['userData', 'followers']),
  );

export {
  selectGlobal,
  makeSelectCurrentUser,
  makeSelectLoading,
  makeSelectError,
  makeSelectRepos,
  makeSelectLocation,
  makeSelectUserProfile,
  makeSelectUserFollowers,
};
