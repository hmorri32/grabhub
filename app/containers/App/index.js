/**
 *
 * App
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 */

import React from 'react';
import { Helmet } from 'react-helmet';
import styled from 'styled-components';
import { Switch, Route } from 'react-router-dom';

import HomePage from 'containers/HomePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import Header from 'components/Header';
import Footer from 'components/Footer';
import Instructions from './Instructions';
import UserProfile from '../UserProfile/Loadable';

import GlobalStyle from '../../global-styles';

const AppWrapper = styled.div`
  max-width: calc(1200px + 16px * 2);
  margin: 0 auto;
  display: flex;
  min-height: 100%;
  padding: 0 16px;
  flex-direction: column;
`;

export default function App() {
  return (
    <AppWrapper>
      <Helmet titleTemplate="HBM Grabhub" defaultTitle="HBM Grabhub">
        <meta name="description" />
      </Helmet>
      <Header />
      <Switch>
        <Route exact path="/" component={HomePage} />
        <Route exact path="/instructions" component={Instructions} />
        <Route exact path="/userProfile" component={UserProfile} />
        <Route path="" component={NotFoundPage} />
      </Switch>
      <Footer />
      <GlobalStyle />
    </AppWrapper>
  );
}
