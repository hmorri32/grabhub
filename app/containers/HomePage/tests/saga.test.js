/**
 * Test: HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';
import { LOAD_REPOS, LOAD_USER_PROFILE } from 'containers/App/constants';
import {
  reposLoaded,
  repoLoadingError,
  userProfileLoaded,
  userProfileLoadingError,
} from 'containers/App/actions';
import githubData, { getRepos, getUserProfile } from '../saga';
const username = 'mxstbr';

/* eslint-disable redux-saga/yield-effects */
describe('application saga', () => {
  describe('getRepos Saga', () => {
    let getReposGenerator;

    // We have to test twice, once for a successful load and once for an unsuccessful one
    // so we do all the stuff that happens beforehand automatically in the beforeEach
    beforeEach(() => {
      getReposGenerator = getRepos();

      const selectDescriptor = getReposGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();

      const callDescriptor = getReposGenerator.next(username).value;
      expect(callDescriptor).toMatchSnapshot();
    });

    it('should dispatch the reposLoaded action if it requests the data successfully', () => {
      const response = [
        {
          name: 'First repo',
        },
        {
          name: 'Second repo',
        },
      ];
      const putDescriptor = getReposGenerator.next(response).value;
      expect(putDescriptor).toEqual(put(reposLoaded(response, username)));
    });

    it('should call the repoLoadingError action if the response errors', () => {
      const response = new Error('oh no');
      const putDescriptor = getReposGenerator.throw(response).value;
      expect(putDescriptor).toEqual(put(repoLoadingError(response)));
    });
  });

  describe('getUserProfile Saga', () => {
    let getUserProfileGenerator;
    beforeEach(() => {
      getUserProfileGenerator = getUserProfile();

      const selectDescriptor = getUserProfileGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();

      const callDescriptor = getUserProfileGenerator.next(username).value;
      expect(callDescriptor).toMatchSnapshot();
    });

    it('should dispatch the userProfileLoaded action if it requests the data successfully', () => {
      const response = [
        {
          profile: {},
        },
      ];
      const putDescriptor = getUserProfileGenerator.next(response).value;
      expect(putDescriptor).toEqual(put(userProfileLoaded(response)));
    });

    it('should call the userProfileLoadingError action if the response errors', () => {
      const response = new Error('oh no');
      const putDescriptor = getUserProfileGenerator.throw(response).value;
      expect(putDescriptor).toEqual(put(userProfileLoadingError(response)));
    });
  });

  describe('githubDataSaga Saga', () => {
    const githubDataSaga = githubData();

    it('should start task to watch for LOAD_REPOS action', () => {
      const takeLatestDescriptor = githubDataSaga.next().value;
      expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_REPOS, getRepos));
    });

    it('should start task to watch for LOAD_USER_PROFILE action', () => {
      const takeLatestDescriptor = githubDataSaga.next().value;
      expect(takeLatestDescriptor).toEqual(
        takeLatest(LOAD_USER_PROFILE, getUserProfile),
      );
    });
  });
});
