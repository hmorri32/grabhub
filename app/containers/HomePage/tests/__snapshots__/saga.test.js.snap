// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application saga getRepos Saga should call the repoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`application saga getRepos Saga should call the repoLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`application saga getRepos Saga should dispatch the reposLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`application saga getRepos Saga should dispatch the reposLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`application saga getUserFollowers Saga should call the userProfileLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`application saga getUserFollowers Saga should call the userProfileLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/followers",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`application saga getUserFollowers Saga should dispatch the userFollowersLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`application saga getUserFollowers Saga should dispatch the userFollowersLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/followers",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`application saga getUserProfile Saga should call the userProfileLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`application saga getUserProfile Saga should call the userProfileLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`application saga getUserProfile Saga should dispatch the userProfileLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`application saga getUserProfile Saga should dispatch the userProfileLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
